### ISSUE TITLE
AvroSerializer.WriterSchema can include invalid unions when there are multiple known types that match a generic IEnumerable

### ISSUE THREAD

**Issue by sokhalsa** [Azure/azure-sdk-for-net#989]
According to the Avro spec: "Unions may not contain more than one schema with the same type, except for the named types record, fixed and enum."

Given a class containing an IEnumerable:

```
[DataContract]
internal class IEnumerableClass<T> : IEquatable<IEnumerableClass<T>>
{
    public static IEnumerableClass<T> Create(IEnumerable<T> enumerable)
    {
        return new IEnumerableClass<T> { Property = enumerable };
    }

    [DataMember]
    public IEnumerable<T> Property { get; set; }

    public bool Equals(IEnumerableClass<T> other)
    {
        if (other == null)
        {
            return false;
        }

        if (this.Property.Count() != other.Property.Count())
        {
            return false;
        }

        return this.Property.Zip(other.Property, (a, b) => new Tuple<T, T>(a, b))
            .ToList()
            .TrueForAll(tuple => tuple.Item1.Equals(tuple.Item2));
    }

    public override bool Equals(object obj)
    {
        return this.Equals(obj as IEnumerableClass<T>);
    }

    public override int GetHashCode()
    {
        return base.GetHashCode();
    }
}
```

If there are multiple known types specified that are a match for the IEnumerable, the writer schema that is generated will contain a union with multiple array schemas. Using the generated writer schema in a call to AvroSerializer.CreateDeserializer<T> results in a SerializationException ("Unions cannot contains schemas of the same type: 'array'").

The following unit test demonstrates the issue:

```
[TestMethod]
[TestCategory("CheckIn")]
public void Serializer_SerializeClassWithGenericMemberHavingMultipleMatchingKnownTypes()
{
    var expected = IEnumerableClass<string>.Create(new List<string> { "aaa", "bbb", "ccc" });

    Type type = expected.GetType();
    Type[] typeArguments = type.GetGenericArguments();

    var settings = new AvroSerializerSettings
    {
        Resolver = new AvroDataContractResolver(),
        KnownTypes = new List<Type>
        {
            typeof(List<>).MakeGenericType(typeArguments),
            typeArguments[0].MakeArrayType()
        }
    };

    var serializer = AvroSerializer.Create<IEnumerableClass<string>>(settings);
    var deserializer = AvroSerializer.CreateDeserializerOnly<IEnumerableClass<string>>(serializer.WriterSchema.ToString(), settings);

    using (var stream = new MemoryStream())
    {
        serializer.Serialize(stream, expected);
        stream.Seek(0, SeekOrigin.Begin);
        var actual = deserializer.Deserialize(stream);

        Assert.AreEqual(expected, actual);
    }
}
```

**Comment by chuanboz** [comment]
proposed following fix for #989
https://github.com/chuanboz/HDInsight/commit/bea99186080ac0a3539fe2cd1b3d6bc4cec1a41d

**Comment by shahabhijeet** [comment]
@sokhalsa if this is still an issue, please open a new issue to track it.
